providers/rpc/rpc.go

```diff
@@ -0,0 +1,351 @@
+	if err != nil {
+		return err
+	}
+	// test that we can communicate with the rpc consumer.
+	// and that it responses with the spec contract (Response{}).
+	if _, err := c.Opts.Request.Client.Do(testReq); err != nil { //nolint:bodyclose // not reading the body
+		return err
+	}
+
+	return nil
+}
+
+// Close a connection to the rpc consumer.
+func (c *Config) Close(_ context.Context) (err error) {
+	return nil
+}
+
+// BootDeviceSet sends a next boot device rpc notification.
+func (c *Config) BootDeviceSet(ctx context.Context, bootDevice string, setPersistent, efiBoot bool) (ok bool, err error) {
+	p := RequestPayload{
+		ID:     int64(time.Now().UnixNano()),
+		Host:   c.Host,
+		Method: BootDeviceMethod,
+		Params: BootDeviceParams{
+			Device:     bootDevice,
+			Persistent: setPersistent,
+			EFIBoot:    efiBoot,
+		},
+	}
+	rp, err := c.process(ctx, p)
+	if err != nil {
+		return false, err
+	}
+	if rp.Error != nil {
+		return false, fmt.Errorf("error from rpc consumer: %v", rp.Error)
+	}
+
+	return true, nil
+}
+
+// PowerSet sets the power state of a BMC machine.
+func (c *Config) PowerSet(ctx context.Context, state string) (ok bool, err error) {
+	switch strings.ToLower(state) {
+	case "on", "off", "cycle":
+		p := RequestPayload{
+			ID:     int64(time.Now().UnixNano()),
+			Host:   c.Host,
+			Method: PowerSetMethod,
+			Params: PowerSetParams{
+				State: strings.ToLower(state),
+			},
+		}
+		resp, err := c.process(ctx, p)
+		if err != nil {
+			return ok, err
+		}
+		if resp.Error != nil {
+			return ok, fmt.Errorf("error from rpc consumer: %v", resp.Error)
+		}
+
+		return true, nil
+	}
+
+	return false, errors.New("requested power state is not supported")
+}
+
+// PowerStateGet gets the power state of a BMC machine.
+func (c *Config) PowerStateGet(ctx context.Context) (state string, err error) {
+	p := RequestPayload{
+		ID:     int64(time.Now().UnixNano()),
+		Host:   c.Host,
+		Method: PowerGetMethod,
+	}
+	resp, err := c.process(ctx, p)
+	if err != nil {
+		return "", err
+	}
+	if resp.Error != nil {
+		return "", fmt.Errorf("error from rpc consumer: %v", resp.Error)
+	}
+
+	return resp.Result.(string), nil
+}
+
+// process is the main function for the roundtrip of rpc calls to the ConsumerURL.
+func (c *Config) process(ctx context.Context, p RequestPayload) (ResponsePayload, error) {
+	// 1. create the HTTP request.
+	// 2. create the signature payload.
+	// 3. sign the signature payload.
+	// 4. add signatures to the request as headers.
+	// 5. request/response round trip.
+	// 6. handle the response.
+	req, err := c.createRequest(ctx, p)
+	if err != nil {
+		return ResponsePayload{}, err
+	}
+
+	// create the signature payload
+	// get the body and reset it as readers can only be read once.
+	body, err := io.ReadAll(req.Body)
+	if err != nil {
+		return ResponsePayload{}, err
+	}
+	req.Body = io.NopCloser(bytes.NewBuffer(body))
```

Comment:
Similar to my comment in the `requestKVS` function, I think this is a leak.

Reply:
sorry, dont know if i follow. mind elaborating?